File: packages/deon-java/source/harness/Harness.java
package harness;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.PrintStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import deon.Deon;
import deon.DeonException;
import deon.DeonMap;
import deon.Diagnostic;
import deon.Entity;
import deon.ParseOptions;
import deon.StringifyOptions;

/**
 * The cross-implementation harness adapter (spec/harness/README.md). A filter: newline-delimited JSON
 * in, newline-delimited JSON out. Every value in a request and a response is a string, so the request
 * itself parses with the implementation's own JSON reader and no third-party decoder is needed.
 */
public final class Harness {
    public static void main(String[] args) throws Exception {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8));
        PrintStream out = new PrintStream(System.out, true, StandardCharsets.UTF_8);
        String line;
        while ((line = reader.readLine()) != null) {
            if (line.isEmpty()) {
                continue;
            }
            DeonMap request;
            try {
                Object parsed = Deon.readJson(line, "<request>");
                if (!(parsed instanceof DeonMap map)) {
                    continue;
                }
                request = map;
            } catch (DeonException e) {
                continue;
            }
            out.println(answer(request));
        }
    }

    private static String answer(DeonMap request) {
        String id = str(request, "id", "");
        try {
            return perform(request, id);
        } catch (DeonException e) {
            Diagnostic d = e.primary();
            StringBuilder b = new StringBuilder();
            b.append("{\"id\":");
            jsonString(b, id);
            b.append(",\"ok\":\"false\",\"code\":");
            jsonString(b, e.code.wire());
            b.append(",\"line\":");
            jsonString(b, Integer.toString(d.span.line));
            b.append(",\"column\":");
            jsonString(b, Integer.toString(d.span.column));
            b.append('}');
            return b.toString();
        } catch (RuntimeException e) {
            return "{\"id\":" + quoted(id) + ",\"ok\":\"false\",\"code\":\"HOST_PANIC\",\"line\":\"0\",\"column\":\"0\"}";
        }
    }

    private static String perform(DeonMap request, String id) {
        String op = str(request, "op", "");
        String source = str(request, "source", "");
        String sourceName = str(request, "sourceName", "<memory>");

        if (op.equals("entities")) {
            StringBuilder j = new StringBuilder("[");
            List<Entity> entities = Deon.entities(source, sourceName);
            for (int i = 0; i < entities.size(); i++) {
                if (i > 0) {
                    j.append(',');
                }
                Entity e = entities.get(i);
                j.append("{\"name\":");
                jsonString(j, e.name());
                j.append(",\"parameters\":[");
                for (int p = 0; p < e.parameters().size(); p++) {
                    if (p > 0) {
                        j.append(',');
                    }
                    jsonString(j, e.parameters().get(p));
                }
                j.append("],\"kind\":");
                jsonString(j, e.kind());
                j.append('}');
            }
            j.append(']');
            return ok(id, j.toString());
        }

        if (op.equals("lint")) {
            StringBuilder j = new StringBuilder("[");
            List<Diagnostic> diagnostics = Deon.lint(source, sourceName);
            for (int i = 0; i < diagnostics.size(); i++) {
                if (i > 0) {
                    j.append(',');
                }
                Diagnostic d = diagnostics.get(i);
                j.append("{\"code\":");
                jsonString(j, d.code.wire());
                j.append(",\"line\":");
                jsonString(j, Integer.toString(d.span.line));
                j.append(",\"column\":");
                jsonString(j, Integer.toString(d.span.column));
                j.append('}');
            }
            j.append(']');
            return ok(id, j.toString());
        }

        Object value = Deon.parseWith(source, optionsOf(request));
        return switch (op) {
            case "canonical" -> ok(id, Deon.canonical(value));
            case "stringify" -> ok(id, Deon.stringify(value, stringifyOptionsOf(request)));
            case "typed" -> ok(id, marshal(Deon.typed(value)));
            case "datasign" -> ok(id, marshal(value)); // parseWith already applied the contracts
            default -> "{\"id\":" + quoted(id) + ",\"ok\":\"false\",\"code\":\"HOST_PANIC\",\"line\":\"0\",\"column\":\"0\"}";
        };
    }

    private static String ok(String id, String result) {
        StringBuilder b = new StringBuilder();
        b.append("{\"id\":");
        jsonString(b, id);
        b.append(",\"ok\":\"true\",\"result\":");
        jsonString(b, result);
        b.append('}');
        return b.toString();
    }

    // #region request access
    private static ParseOptions optionsOf(DeonMap request) {
        ParseOptions o = new ParseOptions();
        o.sourceName = str(request, "sourceName", "<memory>");
        o.filebase = str(request, "filebase", "");
        o.resources = stringMap(request, "files");
        o.absolutePaths = stringMap(request, "absolutePaths");
        o.environment = stringMap(request, "environment");
        o.allowFilesystem = str(request, "allowFilesystem", "false").equals("true");
        o.allowNetwork = str(request, "allowNetwork", "false").equals("true");
        o.datasignFiles = stringList(request, "datasignFiles");
        o.datasignMap = stringMap(request, "datasignMap");
        return o;
    }

    private static StringifyOptions stringifyOptionsOf(DeonMap request) {
        StringifyOptions o = StringifyOptions.defaults();
        Object raw = request.get("stringifyOptions");
        if (!(raw instanceof DeonMap m)) {
            return o;
        }
        o.canonical = truthy(m, "canonical", false);
        o.readable = truthy(m, "readable", true);
        o.leaflinks = truthy(m, "leaflinks", false);
        o.leaflinkShortening = truthy(m, "leaflinkShortening", true);
        o.generatedHeader = truthy(m, "generatedHeader", false);
        o.generatedComments = truthy(m, "generatedComments", false);
        o.indentation = number(m, "indentation", 4);
        o.leaflinkLevel = number(m, "leaflinkLevel", 1);
        return o;
    }

    private static boolean truthy(DeonMap m, String key, boolean fallback) {
        Object v = m.get(key);
        return v instanceof String s ? s.equals("true") : fallback;
    }

    private static int number(DeonMap m, String key, int fallback) {
        Object v = m.get(key);
        if (v instanceof String s) {
            try {
                return Integer.parseInt(s);
            } catch (NumberFormatException ignored) {
                return fallback;
            }
        }
        return fallback;
    }

    private static String str(DeonMap m, String key, String fallback) {
        Object v = m.get(key);
        return v instanceof String s ? s : fallback;
    }

    private static Map<String, String> stringMap(DeonMap m, String key) {
        Object v = m.get(key);
        if (!(v instanceof DeonMap nested)) {
            return Map.of();
        }
        Map<String, String> out = new LinkedHashMap<>();
        for (String k : nested.keys()) {
            Object item = nested.get(k);
            out.put(k, item instanceof String s ? s : "");
        }
        return out;
    }

    private static List<String> stringList(DeonMap m, String key) {
        Object v = m.get(key);
        if (!(v instanceof List<?> list)) {
            return List.of();
        }
        List<String> out = new ArrayList<>();
        for (Object item : list) {
            out.add(item instanceof String s ? s : "");
        }
        return out;
    }
    // #endregion

    // #region JSON output
    private static String marshal(Object v) {
        StringBuilder b = new StringBuilder();
        marshal(b, v);
        return b.toString();
    }

    private static void marshal(StringBuilder b, Object v) {
        if (v instanceof String s) {
            jsonString(b, s);
        } else if (v instanceof Boolean bool) {
            b.append(bool ? "true" : "false");
        } else if (v instanceof Double d) {
            // A whole number within signed 64-bit range writes as a bare integer, matching the sibling
            // adapters. Anything larger must not narrow through a long: (long) 1e19 saturates to
            // Long.MAX_VALUE and would hand back a different number than the one typed (spec 14). Such a
            // value, and every fractional one, writes through its shortest round-tripping decimal.
            if (d == Math.rint(d) && Math.abs(d) < 0x1p63) {
                b.append(Long.toString(d.longValue()));
            } else {
                b.append(d.toString());
            }
        } else if (v instanceof List<?> list) {
            b.append('[');
            for (int i = 0; i < list.size(); i++) {
                if (i > 0) {
                    b.append(',');
                }
                marshal(b, list.get(i));
            }
            b.append(']');
        } else if (v instanceof DeonMap map) {
            b.append('{');
            List<String> keys = map.keys();
            for (int i = 0; i < keys.size(); i++) {
                if (i > 0) {
                    b.append(',');
                }
                jsonString(b, keys.get(i));
                b.append(':');
                marshal(b, map.get(keys.get(i)));
            }
            b.append('}');
        } else {
            b.append("null");
        }
    }

    private static String quoted(String s) {
        StringBuilder b = new StringBuilder();
        jsonString(b, s);
        return b.toString();
    }

    private static void jsonString(StringBuilder b, String s) {
        b.append('"');
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '"' -> b.append("\\\"");
                case '\\' -> b.append("\\\\");
                case '\n' -> b.append("\\n");
                case '\r' -> b.append("\\r");
                case '\t' -> b.append("\\t");
                case '\b' -> b.append("\\b");
                case '\f' -> b.append("\\f");
                default -> {
                    if (c < 0x20) {
                        b.append(String.format("\\u%04x", (int) c));
                    } else {
                        b.append(c);
                    }
                }
            }
        }
        b.append('"');
    }
    // #endregion
}

File: packages/deon-rust/src/bin/harness.rs
//! The cross-implementation harness adapter (`spec/harness/README.md`).
//!
//! A filter: newline-delimited JSON in, newline-delimited JSON out.
//!
//! It uses Deon's *own* JSON reader and writer, so it needs no dependency to take part — which is
//! the reason the protocol carries every value as a string. A Deon value is a string (specification
//! 2), so a request decoded by `deon::json::parse_json` is already exactly what an adapter needs,
//! and nobody had to agree on a decoder first.

use std::io::{self, BufRead, Write};

use deon::json::{parse_json, write_json, write_typed_json};
use deon::value::{Map, Value};
use deon::{ParseOptions, StringifyOptions};

fn text<'a>(request: &'a Map, key: &str) -> &'a str {
    match request.get(key) {
        Some(Value::String(value)) => value,
        _ => "",
    }
}

fn flag(request: &Map, key: &str, fallback: bool) -> bool {
    match request.get(key) {
        Some(Value::String(value)) => value == "true",
        _ => fallback,
    }
}

fn number(request: &Map, key: &str, fallback: usize) -> usize {
    match request.get(key) {
        Some(Value::String(value)) => value.parse().unwrap_or(fallback),
        _ => fallback,
    }
}

fn table<'a>(request: &'a Map, key: &str) -> Vec<(&'a str, &'a str)> {
    match request.get(key) {
        Some(Value::Map(entries)) => entries
            .iter()
            .filter_map(|(name, value)| match value {
                Value::String(item) => Some((name.as_str(), item.as_str())),
                _ => None,
            })
            .collect(),
        _ => Vec::new(),
    }
}

fn options_of(request: &Map) -> ParseOptions {
    let mut options = ParseOptions::new()
        .source_name(text(request, "sourceName"))
        .filebase(text(request, "filebase"))
        .allow_filesystem(flag(request, "allowFilesystem", false))
        .allow_network(flag(request, "allowNetwork", false));

    if text(request, "sourceName").is_empty() {
        options = options.source_name("<memory>");
    }

    for (target, data) in table(request, "files") {
        options = options.resource(target, data);
    }

    for (from, to) in table(request, "absolutePaths") {
        options = options.absolute_path(from, to);
    }

    for (name, value) in table(request, "environment") {
        options = options.environment_variable(name, value);
    }

    // The contracts of specification 14.1, when the request carries any. The files themselves arrive
    // through `files`, like every other resource, so no adapter reaches a disk.
    if let Some(Value::List(files)) = request.get("datasignFiles") {
        options.datasign_files = files
            .iter()
            .filter_map(|file| file.as_str().map(str::to_string))
            .collect();
    }

    for (key, declared) in table(request, "datasignMap") {
        options
            .datasign_map
            .insert(key.to_string(), declared.to_string());
    }

    options
}

fn stringify_options_of(request: &Map) -> StringifyOptions {
    let given = match request.get("stringifyOptions") {
        Some(Value::Map(entries)) => entries.clone(),
        _ => Map::new(),
    };

    StringifyOptions {
        canonical: flag(&given, "canonical", false),
        readable: flag(&given, "readable", true),
        indentation: number(&given, "indentation", 4),
        leaflinks: flag(&given, "leaflinks", false),
        leaflink_level: number(&given, "leaflinkLevel", 1),
        leaflink_shortening: flag(&given, "leaflinkShortening", true),
        generated_header: flag(&given, "generatedHeader", false),
        generated_comments: flag(&given, "generatedComments", false),
    }
}

fn kind_of(kind: deon::EntityKind) -> &'static str {
    match kind {
        deon::EntityKind::Scalar => "scalar",
        deon::EntityKind::Map => "map",
        deon::EntityKind::List => "list",
        deon::EntityKind::Structure => "structure",
        deon::EntityKind::Link => "link",
        deon::EntityKind::Call => "call",
        deon::EntityKind::Resource => "resource",
    }
}

fn run(request: &Map) -> deon::DResult<String> {
    let operation = text(request, "op");
    let source = text(request, "source");
    let options = options_of(request);

    match operation {
        "entities" => {
            let found = deon::entities(source, &options.source_name)?;

            let list = Value::List(
                found
                    .into_iter()
                    .map(|entity| {
                        let mut map = Map::new();

                        map.insert("name", Value::String(entity.name));
                        map.insert(
                            "parameters",
                            Value::List(entity.parameters.into_iter().map(Value::String).collect()),
                        );
                        map.insert("kind", Value::String(kind_of(entity.kind).into()));

                        Value::Map(map)
                    })
                    .collect(),
            );

            Ok(write_json(&list))
        }
        "lint" => {
            let diagnostics = deon::lint(source, "<memory>")?;

            let list = Value::List(
                diagnostics
                    .into_iter()
                    .map(|diagnostic| {
                        let mut map = Map::new();

                        map.insert("code", Value::String(diagnostic.code.to_string()));
                        map.insert("line", Value::String(diagnostic.span.line.to_string()));
                        map.insert(
                            "column",
                            Value::String(diagnostic.span.column.to_string()),
                        );

                        Value::Map(map)
                    })
                    .collect(),
            );

            Ok(write_json(&list))
        }
        _ => {
            let value = deon::parse_with(source, &options)?;

            match operation {
                "canonical" => deon::canonical(&value),
                "stringify" => deon::stringify(&value, &stringify_options_of(request)),
                "typed" => Ok(write_typed_json(&deon::typed(&value)?)),
                "datasign" => Ok(write_typed_json(&deon::sign(&value, &options)?)),
                other => panic!("unknown operation '{other}'"),
            }
        }
    }
}

/// A JSON string, on one line.
///
/// The crate's own `write_json` indents, which is the right thing for a document and the wrong thing
/// for a protocol that is one response per line. The envelope is therefore written by hand; the
/// *payload* is still whatever the crate produced, carried across as a string.
fn quote(text: &str) -> String {
    let mut out = String::with_capacity(text.len() + 2);

    out.push('"');

    for character in text.chars() {
        match character {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            c if (c as u32) < 0x20 => out.push_str(&format!("\\u{:04x}", c as u32)),
            c => out.push(c),
        }
    }

    out.push('"');

    out
}

fn answer(id: &str, result: deon::DResult<String>) -> String {
    match result {
        Ok(text) => format!(
            "{{{}:{},{}:{},{}:{}}}",
            quote("id"),
            quote(id),
            quote("ok"),
            quote("true"),
            quote("result"),
            quote(&text),
        ),
        Err(failure) => {
            let diagnostic = &failure.diagnostics[0];
            let span = &diagnostic.span;

            // The secondary positions, each a [start, line, column] triple of strings, exactly the
            // three measures the primary carries. Built by hand like the rest of the envelope, so a
            // diagnostic with nothing else to point at emits `[]`.
            let related = format!(
                "[{}]",
                diagnostic
                    .related
                    .iter()
                    .map(|related_span| format!(
                        "[{},{},{}]",
                        quote(&related_span.start.to_string()),
                        quote(&related_span.line.to_string()),
                        quote(&related_span.column.to_string()),
                    ))
                    .collect::<Vec<_>>()
                    .join(",")
            );

            format!(
                "{{{}:{},{}:{},{}:{},{}:{},{}:{},{}:{},{}:{},{}:{}}}",
                quote("id"),
                quote(id),
                quote("ok"),
                quote("false"),
                quote("code"),
                quote(&failure.code.to_string()),
                quote("severity"),
                quote("error"),
                quote("start"),
                quote(&span.start.to_string()),
                quote("line"),
                quote(&span.line.to_string()),
                quote("column"),
                quote(&span.column.to_string()),
                quote("related"),
                related,
            )
        }
    }
}

fn main() {
    let stdin = io::stdin();
    let mut stdout = io::stdout();

    for line in stdin.lock().lines() {
        let line = line.expect("stdin should be readable");

        if line.trim().is_empty() {
            continue;
        }

        let request = match parse_json(&line) {
            Ok(Value::Map(map)) => map,
            _ => panic!("a request must be a JSON object"),
        };

        let id = text(&request, "id").to_string();

        writeln!(stdout, "{}", answer(&id, run(&request))).expect("stdout should be writable");
        stdout.flush().expect("stdout should flush");
    }
}
